doc/source/dev/contributor/devpy_test.rst

```diff
@@ -120,6 +120,14 @@ which are disabled even when calling ``python dev.py test -m full``.
 They can be enabled by setting the environment variable ``SCIPY_XSLOW=1``
 before running the test suite.
 
+By default, tests that use ``Hypothesis`` run with the ``deterministic``
+profile defined in ``scipy/scipy/conftest.py``. This profile includes the
+Hypothesis setting ``derandomize=True`` so the same examples are used until
+Hypothesis, Python, or the test function are updated. To better use
```

Comment:
It doesn't look like you're pinning Hypothesis (or any other dependencies) when installing in CI, so there will probably be some occasions on which we release a new version - the exact generated examples are explicitly unstable and can change at any time - and the first person to run into that is a new contributor rather than a maintainer.

IMO _just_ pinning Hypothesis would do more harm than that by missing out on new features/bugfixes/perf, but I'd recommend an e.g. `pip-compile` based workflow to pin your transitive dependencies and update them regularly - Hypothesis itself does this with a weekly cronjob.

Reply:
> IMO just pinning Hypothesis would do more harm than that by missing out on new features/bugfixes/perf

Agreed I think

>  the exact generated examples are explicitly unstable and can change at any time 

How often does this happen? Every minor/bugfix release, or only on occasion?

> workflow to pin your transitive dependencies and update them regularly

This is a big job to do well, so I'd prefer not to couple it to this PR unless we really must. And if we do it then I'd much rather go the `conda-lock` route like scikit-learn does, so we can properly control all our dependencies (e.g., testing different BLAS/LAPACK libraries rather than the exact same build of OpenBLAS over and over).

Reply:
Thoroughly irregular, but there are changes in the underlying engine a few times per year and a few changes to one or more particular strategies most months.  Of course, any given change is pretty unlikely to make a given test fail!

Agree that pinning deps is a big job and best separated; just wanted to flag it here because when pinned Hypothesis really truly is deterministic.

Reply:
Thanks @Zac-HD. I think we're fine with leaving it unpinned for now then. Given we have very few tests and the chance of breakage is low (it a single change, not a "now things are fully non-deterministic"), we should be fine for quite a while. It's similar to other unpinned deps that may break CI jobs with new releases.